File: benchmarks/drivers/go-mysql/mix.go
package main

import (
	"math/rand"
	"os"
	"strconv"
)

// kWarehouses mirrors the Dart bench's TPCC_WAREHOUSES env knob so all
// drivers in a session agree on the scale.
var kWarehouses = envInt("TPCC_WAREHOUSES", 3)

const (
	kDistrictsPerWarehouse = 10
	kCustomersPerDistrict  = 300
	kItems                 = 100000
)

func envInt(k string, def int) int {
	if v := os.Getenv(k); v != "" {
		if n, err := strconv.Atoi(v); err == nil {
			return n
		}
	}
	return def
}

type TxType int

const (
	TxNewOrder TxType = iota
	TxPayment
	TxOrderStatus
	TxDelivery
	TxStockLevel
)

func (t TxType) Name() string {
	switch t {
	case TxNewOrder:
		return "newOrder"
	case TxPayment:
		return "payment"
	case TxOrderStatus:
		return "orderStatus"
	case TxDelivery:
		return "delivery"
	case TxStockLevel:
		return "stockLevel"
	}
	return "unknown"
}

type NewOrderLine struct {
	IID       int
	SupplyWID int
	Quantity  int
}

type NewOrderParams struct {
	WID, DID, CID int
	Lines         []NewOrderLine
}

type PaymentParams struct {
	WID, DID, CID int
	Amount        float64
}

type OrderStatusParams struct{ WID, DID, CID int }
type DeliveryParams struct{ WID, CarrierID int }
type StockLevelParams struct{ WID, DID, Threshold int }

// TpccMix mirrors `benchmarks/scripts/tpcc/mix.dart`.
type TpccMix struct {
	rng *rand.Rand
}

func NewTpccMix(seed int64) *TpccMix {
	return &TpccMix{rng: rand.New(rand.NewSource(seed))}
}

func (m *TpccMix) NextType() TxType {
	r := m.rng.Intn(100)
	switch {
	case r < 45:
		return TxNewOrder
	case r < 88:
		return TxPayment
	case r < 92:
		return TxOrderStatus
	case r < 96:
		return TxDelivery
	default:
		return TxStockLevel
	}
}

func (m *TpccMix) w() int    { return 1 + m.rng.Intn(kWarehouses) }
func (m *TpccMix) d() int    { return 1 + m.rng.Intn(kDistrictsPerWarehouse) }
func (m *TpccMix) c() int    { return 1 + m.rng.Intn(kCustomersPerDistrict) }
func (m *TpccMix) item() int { return 1 + m.rng.Intn(kItems) }

func (m *TpccMix) NewOrder() NewOrderParams {
	wID := m.w()
	olCnt := 5 + m.rng.Intn(11)
	lines := make([]NewOrderLine, 0, olCnt)
	for i := 0; i < olCnt; i++ {
		remote := m.rng.Intn(100) == 0
		supply := wID
		if remote {
			supply = 1 + m.rng.Intn(kWarehouses)
		}
		lines = append(lines, NewOrderLine{
			IID:       m.item(),
			SupplyWID: supply,
			Quantity:  1 + m.rng.Intn(10),
		})
	}
	return NewOrderParams{WID: wID, DID: m.d(), CID: m.c(), Lines: lines}
}

func (m *TpccMix) Payment() PaymentParams {
	wID := m.w()
	amount := 1.0 + float64(m.rng.Intn(5000)) + m.rng.Float64()
	amount = float64(int64(amount*100+0.5)) / 100.0
	return PaymentParams{WID: wID, DID: m.d(), CID: m.c(), Amount: amount}
}

func (m *TpccMix) OrderStatus() OrderStatusParams {
	return OrderStatusParams{WID: m.w(), DID: m.d(), CID: m.c()}
}

func (m *TpccMix) Delivery() DeliveryParams {
	return DeliveryParams{WID: m.w(), CarrierID: 1 + m.rng.Intn(10)}
}

func (m *TpccMix) StockLevel() StockLevelParams {
	return StockLevelParams{WID: m.w(), DID: m.d(), Threshold: 10 + m.rng.Intn(11)}
}

File: benchmarks/drivers/java/src/main/java/io/github/tganzerli/bench/Mix.java
package io.github.tganzerli.bench;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Mix {
    public static final int K_WAREHOUSES = 3;
    public static final int K_DISTRICTS = 10;
    public static final int K_CUSTOMERS = 300;
    public static final int K_ITEMS = 100_000;

    public enum TxType {
        NEW_ORDER("newOrder"),
        PAYMENT("payment"),
        ORDER_STATUS("orderStatus"),
        DELIVERY("delivery"),
        STOCK_LEVEL("stockLevel");
        public final String name;
        TxType(String n) { this.name = n; }
    }

    public static class NewOrderLine {
        public final int iId;
        public final int supplyWId;
        public final int quantity;
        public NewOrderLine(int iId, int supplyWId, int quantity) {
            this.iId = iId; this.supplyWId = supplyWId; this.quantity = quantity;
        }
    }
    public static class NewOrderParams {
        public final int wId, dId, cId;
        public final List<NewOrderLine> lines;
        public NewOrderParams(int wId, int dId, int cId, List<NewOrderLine> lines) {
            this.wId = wId; this.dId = dId; this.cId = cId; this.lines = lines;
        }
    }
    public static class PaymentParams {
        public final int wId, dId, cId;
        public final double amount;
        public PaymentParams(int wId, int dId, int cId, double amount) {
            this.wId = wId; this.dId = dId; this.cId = cId; this.amount = amount;
        }
    }
    public static class OrderStatusParams {
        public final int wId, dId, cId;
        public OrderStatusParams(int wId, int dId, int cId) { this.wId = wId; this.dId = dId; this.cId = cId; }
    }
    public static class DeliveryParams {
        public final int wId, carrierId;
        public DeliveryParams(int wId, int carrierId) { this.wId = wId; this.carrierId = carrierId; }
    }
    public static class StockLevelParams {
        public final int wId, dId, threshold;
        public StockLevelParams(int wId, int dId, int threshold) { this.wId = wId; this.dId = dId; this.threshold = threshold; }
    }

    private final Random rng;
    public Mix(long seed) { this.rng = new Random(seed); }

    public TxType nextType() {
        int r = rng.nextInt(100);
        if (r < 45) return TxType.NEW_ORDER;
        if (r < 88) return TxType.PAYMENT;
        if (r < 92) return TxType.ORDER_STATUS;
        if (r < 96) return TxType.DELIVERY;
        return TxType.STOCK_LEVEL;
    }

    private int w() { return 1 + rng.nextInt(K_WAREHOUSES); }
    private int d() { return 1 + rng.nextInt(K_DISTRICTS); }
    private int c() { return 1 + rng.nextInt(K_CUSTOMERS); }
    private int item() { return 1 + rng.nextInt(K_ITEMS); }

    public NewOrderParams newOrder() {
        int wId = w();
        int olCnt = 5 + rng.nextInt(11);
        int dId = d();
        int cId = c();
        List<NewOrderLine> lines = new ArrayList<>(olCnt);
        for (int i = 0; i < olCnt; i++) {
            boolean remote = rng.nextInt(100) == 0;
            int supply = remote ? 1 + rng.nextInt(K_WAREHOUSES) : wId;
            lines.add(new NewOrderLine(item(), supply, 1 + rng.nextInt(10)));
        }
        return new NewOrderParams(wId, dId, cId, lines);
    }

    public PaymentParams payment() {
        int wId = w();
        int dId = d();
        int cId = c();
        double amount = 1.0 + rng.nextInt(5000) + rng.nextDouble();
        amount = Math.round(amount * 100.0) / 100.0;
        return new PaymentParams(wId, dId, cId, amount);
    }

    public OrderStatusParams orderStatus() { return new OrderStatusParams(w(), d(), c()); }
    public DeliveryParams delivery() { return new DeliveryParams(w(), 1 + rng.nextInt(10)); }
    public StockLevelParams stockLevel() { return new StockLevelParams(w(), d(), 10 + rng.nextInt(11)); }
}
